### PR TITLE
Applications controller: handle deploys opts

### LINKED ISSUE #11449: applications: support wait, timeout, and atomic on crd and controller
#11336 has been split into 2 issues for easier review and to show progress. This issue implements the options in our crd and controller.  
prerequisite: #11448

### PR DESCRIPTION
**Pull request by vgramer**
**What this PR does / why we need it**:
This PR extends `applicationDefinition` and `ApplicationInstallation` CRD with `DeployOptions.HelmDeployOptions`. 

`DeployOptions` holds the settings specific to the templating method used to deploy the application. `HelmDeployOptions` allow configuring the equivalent of `wait`, `timeout` and `atomic` options of Helm cli.

The `DeployOptions` defined at `ApplicationDefinition` level are used a default if `DeployOptions` if not defined at `ApplicationInstallation` level.

Moreover, ApplicationInstallation will no more be installed / upgraded if the number of failures (`ApplicationInstllation.Status.Failures` ) exceeds the max number of retires (hardcoded to 5). It reduces the load on the cluster and avoids an infinite loop, that disrupts workload, in case atomic is true.

Finally, ApplicationInstallation's ready condition is set to `unknown` with reason `InstallationInProgress` before starting the installation. 

**Which issue(s) this PR fixes**:

Fixes #11449

**What type of PR is this?**

/kind feature
**Special notes for your reviewer**:
The first commits are refactoring commits. 
The validation and defaulting of DeployOptions will be implemented in a further PR

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
* Extends `applicationDefinition` and `ApplicationInstallation` CRD with `DeployOptions.HelmDeployOptions` to control how applications are deployed with `Helm`.
* ApplicationInstallation: set condition ready to `unknown` with reason `InstallationInProgress` before starting the installation.
* ApplicationInstallation: don't try to install / upgrade the application if the max number of retries is exceeded
```

**Documentation**:

```documentation
TDB
```

### CODE REVIEW

File: pkg/applications/providers/template/helm.go [hunk 1]
```diff
@@ -191,3 +194,20 @@ func getReleaseName(applicationInstallation *appskubermaticv1.ApplicationInstall
 	}
 	return namespacedName
 }
+
+// getDeployOps builds helmclient.DeployOpts from values provided by appInstall or fallback to the values of appDefinition or fallback to the default options.
+// Default options are wait=false that implies timeout=0 and atomic=false.
+func getDeployOps(appDefinition *appskubermaticv1.ApplicationDefinition, appInstall *appskubermaticv1.ApplicationInstallation) (*helmclient.DeployOpts, error) {
```
Comment (SimonTheLeg) [inline pkg/applications/providers/template/helm.go]:
```suggestion
func getDeployOpts(appDefinition *appskubermaticv1.ApplicationDefinition, appInstall *appskubermaticv1.ApplicationInstallation) (*helmclient.DeployOpts, error) {
```

File: pkg/applications/providers/template/helm_integration_test.go [hunk 2]
```diff
@@ -194,24 +197,98 @@ func TestHelmProvider(t *testing.T) {
...
-				_, err := template.InstallOrUpgrade(chartFullPath, app)
+				_, err := template.InstallOrUpgrade(chartFullPath, &appskubermaticv1.ApplicationDefinition{}, app)
 				if err == nil {
 					t.Fatal("install of application should have failed but no error was raised")
 				}
 			},
 		},
+		// these tests ensure deploysOpt defined in CR are correctly passed to the helmClient
+		{
+			name: "application installation should fail when timeout is exceeded (timeout is defined at applincation Installation Level)",
```
Comment (SimonTheLeg) [inline pkg/applications/providers/template/helm_integration_test.go]:
```suggestion
			name: "application installation should fail when timeout is exceeded (timeout is defined at application Installation Level)",
```

**Review by SimonTheLeg — CHANGES_REQUESTED** [review]
looks good to me. Two small typos, but the rest is nice :)

File: pkg/applications/providers/template/helm.go (hunk 1 above)
Comment (vgramer) [reply pkg/applications/providers/template/helm.go]: nice catch

**Review by SimonTheLeg — APPROVED** [review]

**Pull request merged**